Kernel/Process.cpp

```diff
@@ -205,12 +205,59 @@ RefPtr<Process> Process::create_kernel_process(RefPtr<Thread>& first_thread, Str
 
 void Process::protect_data()
 {
-    MM.set_page_writable_direct(VirtualAddress { this }, false);
+    u32 expected = 1 << 1;
+    u32 desired = (1 << 1) | 1;
+    // Least significant bit indicates we're busy protecting/unprotecting
+    for (;;) {
+        if (m_protected_data_refs.compare_exchange_strong(expected, desired, AK::memory_order_relaxed))
+            break;
+
+        Processor::wait_check();
+
+        expected &= ~1;
+        VERIFY(expected != 0); // Someone should always have at least one reference
+
+        if (expected == 1 << 1)
+            desired = (1 << 1) | 1;
+        else
+            desired = expected - (1 << 1);
+    }
+
+    AK::atomic_thread_fence(AK::memory_order_release);
+
+    if (expected == 1 << 1) {
```

Comment:
Can we find a suitable synchronization primitive to do this with some kind of RAII one-liner instead of having to have this huge chunk of code here?

Reply:
I thought about this too, but couldn't really think of anything. Do you have any suggestions? `RefCountedSpinLockOnFirstRef`? ;-)